state-recovery/besu-plugin/src/main/kotlin/linea/staterecovery/plugin/BlockHashLookupWithRecoverySupport.kt

```diff
@@ -0,0 +1,71 @@
+package linea.staterecovery.plugin
+
+import net.consensys.encodeHex
+import net.consensys.minusCoercingUnderflow
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.tuweni.bytes.Bytes32
+import org.hyperledger.besu.datatypes.Hash
+import org.hyperledger.besu.evm.blockhash.BlockHashLookup
+import org.hyperledger.besu.evm.frame.MessageFrame
+import java.util.concurrent.ConcurrentHashMap
+
+class BlockHashLookupWithRecoverySupport(
```

Comment:
Can you explain how the added methods related to recovery? Or `recovery` here just means that it's filling methods are adapted to state recovery usecase?

Reply:
This is because Besu does not save blockHash into the DB. Because in state recovery blocks don't have all data, Besu calculates an hash that is "not the correct one". 
whenever a smart contract uses `blockhash (head - 10)` it would get the wrong hash.

To overcome this, we need a custom hash lookback mechanism. That needs to be populated properly at startup and populated properly on the fly when new block is "recovered" with the block hash from the blob, instea of Besu's calculated one.